### PR TITLE
Allow listening for Hue events

### PR DESCRIPTION
**Pull request by balloob**
This will give an instant state update instead of having to poll.

This is MVP with some basic info for a light.

Test it out with `python3 example.py`. To see debug logs, run `DEBUG=1 python3 example.py`

Thanks to a discovery by @Thomas-Vos

### CODE REVIEW

File: aiohue/lights.py [hunk 1]
```diff
@@ -90,6 +90,25 @@ def colorgamut(self):
...
+            state["xy"] = [color["xy"]["x"], color["xy"]["y"]]
+
+        if ct := update.get("color_temperature"):
+            state["ct"] = ct["mirek"]
+
+        if "on" in update:
+            state["on"] = update["on"]["on"]
+
+        if dimming := update.get("dimming"):
+            state["bri"] = dimming["brightness"]
```
Comment (balloob) [inline aiohue/lights.py:106]: Update events have their own format.

These are the ones that I found so far for lights.

File: aiohue/bridge.py [hunk 2]
```diff
@@ -85,13 +104,73 @@ async def request(self, method, path, json=None, auth=True):
...
+                event = await pending_events.get()
+
+                if event["type"] != "update":
+                    self.logger.debug("Unknown event type: %s", event)
+                    continue
+
+                for update in event["data"]:
+                    item_type = update["id_v1"].split("/", 2)[1]
+
+                    if item_type == "lights":
```
Comment (balloob) [inline aiohue/bridge.py:162]: For first version we're only going to support update events for lights.

File: aiohue/bridge.py [hunk 3]
```diff
@@ -85,13 +104,73 @@ async def request(self, method, path, json=None, auth=True):
...
 
-        except ClientConnectionError:
+        except client_exceptions.ClientConnectionError:
             if self.proto is not None:
                 raise
 
             self.proto = "http"
             return await self.request(method, path, json, auth)
 
+    async def request_2(self, method, path):
```
Comment (balloob) [inline aiohue/bridge.py:114]: I am available for all your naming inquiries.

File: aiohue/bridge.py [hunk 4]
```diff
@@ -85,13 +104,73 @@ async def request(self, method, path, json=None, auth=True):
...
+                    item_type = update["id_v1"].split("/", 2)[1]
+
+                    if item_type == "lights":
+                        obj = self.lights.process_update_event(update)
+                        # if obj is None, we didn't know the object
+                        # We could consider triggering a full refresh
+                        if obj is not None:
+                            yield obj
+
+        except asyncio.CancelledError:
```
Comment (MartinHjelmare) [inline aiohue/bridge.py:169]: Would it be safe to only wrap the `await pending_events.get()` with the try... except block? We would decrease the indentation a bit for readers.

File: aiohue/clip.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        entertainment
+        light
+        bridge_home
+        grouped_light
+        room
+        scene
+        """
+        return await self._request_2("get", "clip/v2/resource")
+
+    @staticmethod
```
Comment (MartinHjelmare) [inline aiohue/clip.py:29]: Maybe make this a classmethod instead since we return an instance of Clip.

Comment (MartinHjelmare) [reply aiohue/clip.py:29]:
```suggestion
    @classmethod
```

File: aiohue/clip.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    async def initialize(request_2):
```
Comment (MartinHjelmare) [inline aiohue/clip.py:30]:
```suggestion
    async def initialize(cls, request_2):
```

File: aiohue/clip.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        raw = await request_2("get", "clip/v2/resource")
+        return Clip(raw, request_2)
```
Comment (MartinHjelmare) [inline aiohue/clip.py:32]:
```suggestion
        return cls(raw, request_2)
```

Comment (MartinHjelmare) [reply aiohue/clip.py:32]: This call doesn't match the signature of the class?

File: aiohue/clip.py (hunk 5 above)
Comment (balloob) [reply aiohue/clip.py:29]: Will remove it, we don't use it anymore.

File: aiohue/clip.py (hunk 7 above)
Comment (balloob) [reply aiohue/clip.py:32]: Yeah, it's old. Removing it.

**Review by MartinHjelmare — APPROVED** [review]
Looks good!

**Comment by balloob** [comment]
I'm still testing the timeouts on my end. Currently running a test without a timeout.

**Comment by spudwebb** [comment]
@balloob , @Thomas-Vos, is this hue events API documented anywhere?

**Comment by Thomas-Vos** [comment]
@spudwebb the new API is not documented anywhere, you just need to try it and see what happens. Philips Hue said on their developer website that they have some "significant API announcements coming up soon".

**Comment by willjgit** [comment]
Does this work with the Hue motion sensor yet?

**Comment by s3frank** [comment]
Hi there,
I was running the fasthue custom integration to have more responsiveness on HA side to event from Hue.
When HA enhanced it's io layer for hue to support push eventing that component was deprecated by the developer as mentioned in this thread.
https://github.com/azogue/fasthue/issues/22

After a few weeks it seems it made it from dev branch to stable and I observed it working on my Tap switches amongst things. But as of late with core-2021.8.8 I seem to be getting about a 5 second response rate on key presses in the worst case and it makes me think polling is back on vs push.

This has been a long "debated" topic on the web, but I just cannot figure out for the life of me how or where these settings are. Like can one set the polling rate statically somewhere? or can one toggle between polling vs pushing?

What is the intended / expected behavior for the official HA hue integration here?
Where is this documented?

I looked at these places:
https://www.home-assistant.io/integrations/hue/

And been try with my limit python knowledge to figure it out from here:
https://github.com/home-assistant-libs/aiohue

Thanks!

**Comment by phillipsawyer** [comment]
I am experiencing the same polling slowness as @s3frank (up to 5s) with Philips Hue dimmer switches (V1 / V2).

It seems like Hue dimmer remotes/motion detectors connected through the Philips Hue Hub are still using local polling and I haven't been able to find anywhere that mentions local push having been implemented either for anything other than lights.

According to https://github.com/Q42/Q42.HueApi/issues/255 it sounds like it will be implemented eventually? That's hopeful.

The https://github.com/azogue/fasthue polling speed-up hack still seems to be the best way to have a reasonable response time for Hue remotes/motion detectors.

Thanks

**Comment by phillipsawyer** [comment]
Hue Remotes seem to work great with local push as of 2021.12

**Pull request merged**